src/dict.c

```diff
@@ -226,6 +227,15 @@ int dictShrinkToFit(dict *d)
     return dictShrink(d, minimal);
 }
 
+/* Expand the table to a proper size to contain the elements. */
+int dictExpandToFit(dict *d) {
```

Comment:
and drop htNeedsShrink and htNeedsExpand...

Reply:
I got a little bit confused after reading #12850, where a lot of APIs are changed. I was thinking of making the APIs of  shrinking and expanding symmetric to each other. I will trying to avoid creating unnecessary functions.

Reply:
Taking "expand" as an example, seems we can do something like this:

```c
void dictExpandIfNeeded(dict *d) {
    /* Incremental rehashing already in progress. Return. */
    if (dictIsRehashing(d)) return;

    /* If the hash table is empty expand it to the initial size. */
    if (DICTHT_SIZE(d->ht_size_exp[0]) == 0) {
        dictExpand(d, DICT_HT_INITIAL_SIZE);
        return;
    }

    /* If we reached the 1:1 ratio, and we are allowed to resize the hash
     * table (global setting) or we should avoid it but the ratio between
     * elements/buckets is over the "safe" threshold, we resize doubling
     * the number of buckets. */
    if ((dict_can_resize == DICT_RESIZE_ENABLE &&
         d->ht_used[0] >= DICTHT_SIZE(d->ht_size_exp[0])) ||
        (dict_can_resize != DICT_RESIZE_FORBID &&
         d->ht_used[0] / DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio))
    {
        if (!dictTypeResizeAllowed(d))
            return;
        dictExpand(d, d->ht_used[0] + 1);
    }
}

/* Expand the hash table if needed */
static void _dictExpandIfNeeded(dict *d)
{
    /* Automatic resizing is disallowed. Return */
    if (d->pauseAutoResize > 0) return;

    dictExpandIfNeeded(d);
}
```

And then `tryResizeHashTables` can directly call `dictExpandIfNeeded`.